docs/efficient_gas_calculation_algorithm.md

```diff
@@ -0,0 +1,156 @@
+# Efficient gas calculation algorithm for EVM
+
+This article describes how to efficiently calculate gas and check stack requirements
+for Ethereum Virtual Machine (EVM) instructions.
+
+
+## Instructions metadata
+
+Let's start by defining some basic and universal instructions' parameters.
+
+1. Base gas cost.
+   
+   This is the static gas cost of instructions. Some instructions may have 
+   additional cost depending on their operand values - these have to be
+   handled individually during the instruction execution.
+   
+2. Stack height requirement.
+
+   This is the minimum stack height (number of items on the stack) 
+   required for the instruction execution.
+   
+3. Stack height change.
+
+   This is difference of the stack height before and after the instruction 
+   execution. Can be negative if the instruction pops more items than pushes.
+   
+Examples:
+
+| opcode  | base gas cost | stack height requirement | stack height change |
+| ------- | ------------- | ------------------------ | ------------------- |
+| ADD     | 3             | 2                        | -1                  |
+| EXP     | 50            | 2                        | -1                  |
+| DUP4    | 3             | 4                        | 1                   |
+| SWAP1   | 3             | 2                        | 0                   |
+| ADDRESS | 2             | 0                        | 1                   |
+| CALL    | 700           | 7                        | -6                  |
+
+
+## Basic blocks
+
+A _basic block_ is a sequence of instructions that are executed "straight-line"
+without being interrupted by jumps. I.e. they are nodes in the _control flow graph_.
```

Comment:
What about calls and e.g. stop?

Reply:
Ah stop is mentioned below, but it would be interesting to know about call, since a call does not really influence the internal control flow.

Reply:
As mentioned in Misc, for calls we only need to know the "current" gas counter, which is not directly available in EVMs following this algorithm. Because of that, in evmone we add back the cost of all instructions after the call.